lib/Shell.js

```diff
@@ -0,0 +1,239 @@
+      throw new PS_PROC_ERROR();
+    });
+
+    // set streams encoding
+    proc.stdin.setDefaultEncoding(options.inputEncoding || 'utf8');
+    proc.stdout.setEncoding(options.outputEncoding || 'utf8');
+    proc.stderr.setEncoding(options.outputEncoding || 'utf8');
+
+    // handle startup errors
+    const psErr = new ShellStreamBuffer();
+    proc.stderr.pipe(psErr);
+
+    proc.stdin.on('error', err => {
+      if (['ECONNRESET', 'EPIPE'].includes(err.code)) {
+        // handle epipe
+        setTimeout(() => {
+          if(psErr.isEmpty()) {
+            throw err; // real epipe
+          }
+        });
+        return; // ignore here so proc.once('close') will handle error
+      }
+      throw err; // fallback
+    });
+
+    proc.once('close', code => {
+      psErr.end();
+      this.emit('end', code);
+      this.verbose && logger.info(`PS process ${this.pid} exited with code ${code}`);
+
+      if(!psErr.isEmpty() && isCriticalPSError(psErr.getContents())) {
+        throw new PS_PROC_ERROR(psErr.getContentsAsString());
+      }
+    });
+
+    this.invoke = this.invoke.bind(this, proc);
+    this.stop = this.stop.bind(this, proc);
+    this.dispose = this.dispose.bind(this, proc);
+
+    this.streams = {
+      stdin: proc.stdin,
+      stdout: proc.stdout,
+      stderr: proc.stderr,
+    };
+  }
+  addCommand(command = '', params = []) {
+    return new Promise(resolve => {
+      if(!(command instanceof PSCommand)) {
+        command = new PSCommand(command);
+      }
+
+      this.commands.push(command);
+
+      if(params && params.length > 0) {
+        logger.warn(`params argument for addCommand method will be deprecated soon. use addParameter or addParameters instead.\n
+        https://rannn505.gitbook.io/node-powershell/shell/addCommand`);
+        return this.addParameters(params);
+      }
+
+      return resolve(this.commands);
+    });
+  }
+  addArgument(argument = '') {
+    if(this.commands.length === 0) {
+      return Promise.reject(new PS_ARG_MISS_ERROR('Commands array is empty. please add at least one command before you use addArgument()'));
+    }
+    const lastCommand = this.commands.pop();
+    lastCommand.addArgument(argument);
+    return this.addCommand(lastCommand);
+  }
+  addParameter(parameter = {}) {
+    if(this.commands.length === 0) {
+      return Promise.reject(new PS_ARG_MISS_ERROR('Commands array is empty. please add at least one command before you use addParameter()'));
+    }
+    const lastCommand = this.commands.pop();
+    lastCommand.addParameter(parameter);
+    return this.addCommand(lastCommand);
+  }
+  addParameters(parameters = []) {
+    if(getType(parameters) !== 'Array') {
+      throw new PS_ARG_TYPE_ERROR('Parameters must be an array');
+    }
+    if(parameters.length === 0) {
+      throw new PS_ARG_MISS_ERROR('Parameters are missing');
+    }
+
+    return Promise.all(parameters.map(this.addParameter, this))
+      .then(() => Promise.resolve(this.commands));
```

Comment:
You never need to `Promise.resolve` in a then - a `then` already `Promise.resolve`s its return value

Reply:
:1st_place_medal: